utils/src/status.rs

```diff
@@ -0,0 +1,455 @@
+    fn bitor(self, rhs: Self) -> Self::Output {
+        match (self, rhs) {
+            (Self::Good, _) | (_, Self::Good) => Self::Good,
+            (Self::Unknown, _) | (_, Self::Unknown) => Self::Unknown,
+            (Self::Bad, Self::Bad) => Self::Bad,
+        }
+    }
+}
+
+// FIXME: a few notes:
+// The Statuses struct below came out less "elegant" than I expected it to be.
+// The idea was that the user would parameterize it with a simple numeric enum that contains
+// indices of "status" fields, and the implementation will then choose the appropriate amount
+// of storage for that amount of fields. Unfortunately, it doesn't seem to be possible
+// in Rust currently, given the information that the enum_iterator::Sequence trait provides
+// (namely, it provides the cardinality of the enum via an associated constant, and it's
+// practically impossible to do any kind of specialization based on it).
+// So, the current implementation requires the user to specify the size of the underlying
+// storage explicitly.
+// The other ugly part is that the proper correspondence between the enum's elements count and
+// the size of the storage is only asserted at runtime currently. But this can be fixed by
+// introducing a custom static_assert macro (the available const_assert won't work, see
+// another FIXME below).
+// Finally, yet another ugly part is that I had to "uglify" certain parts of the implementation
+// in order to be able to declare constants of the Statuses type. I don't like how it came out
+// in the end, so I'll probably abandon this idea (see another FIXME below).
+
+/// A set of [Status]'es.
+///
+/// `NUM_BYTES` specifies the size in bytes of the underlying array (each Status field
+/// requires 2 bits of storage, so the maximum number of fields will be `NUM_BYTES * 4`).
+/// `IndexType` is supposed to be a trivial enum type that will be used to access
+/// the fields. The enum must implement [Copy], [Clone], [num_traits::cast::ToPrimitive]
+/// and (optionally) [enum_iterator::Sequence]. E.g.
+/// ```
+/// // Note that the derive macro that implements `num_traits::cast::ToPrimitive`
+/// // is in a different crate `num_derive`.
+/// use num_derive::ToPrimitive;
+/// use enum_iterator::Sequence;
+/// use utils::status::{Status, Statuses};
+///
+/// #[derive(Copy, Clone, ToPrimitive, Sequence)]
+/// enum MyStatusField {
+///     A,
+///     B,
+/// }
+///
+/// type MyStatuses = Statuses<1, MyStatusField>;
+///
+/// fn do_stuff() {
+///     let mut statuses = MyStatuses::new_unknown();
+///     # assert_eq!(statuses.all_good(), Status::Unknown);
+///     statuses.set(MyStatusField::A, Status::Good);
+///     statuses.set(MyStatusField::B, Status::Bad);
+///     # assert_eq!(statuses.all_good(), Status::Bad);
+/// }
+/// # fn main() {
+/// #   do_stuff();
+/// # }
+/// ```
+#[derive(Debug, Clone, Copy, Encode, Decode)]
+pub struct Statuses<const NUM_BYTES: usize, IndexType>([u8; NUM_BYTES], PhantomData<IndexType>)
```

Comment:
I'm confused... why is this not just an enum?

Also this seems to be belong in chainstate types... not in utils.

Reply:
[discussed] The idea was to be able to have an independent three-state status (Good, Bad, Unknown) for each of the checks performed on a block. But the independent statuses are not needed because checks are always performed in the same order. So, this generic implementation will be removed and a simple enum will be used instead.

P.S. we also discussed at some point that the status enum has to be a bitfield, but it seems that there is no need for it to be one, it can be just a simple sequential enum, e.g.:
```
pub enum BlockValidationStage {
    Initial,
    ParentOk,
    CheckBlockOk,
    FullyChecked,
}

pub struct BlockStatus {
    last_successfull_validation_stage: BlockValidationStage,
}
```
Btw, in bitcoin-core, BlockStatus is not a real bitfield either. Rather, it has a sequential part in its lower bits that represents the block check status (BLOCK_VALID_RESERVED = 1, BLOCK_VALID_TREE = 2, BLOCK_VALID_TRANSACTIONS = 3, BLOCK_VALID_CHAIN = 4, BLOCK_VALID_SCRIPTS = 5) and then it also has a bunch of extra flags in its higher bits. For now, we don't need those extra flags, but if/when we do, we'd rather define a separate bitfield enum for them.